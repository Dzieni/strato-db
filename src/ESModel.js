// Drop-in replacement for JsonModel
// Caveats:
// * `.update()` returns the current object at the time of returning, not the one that was updated
//
// Events all type `es/name` and data `[actionEnum, objOrId, id]`
// The id is assigned by the preprocessor

import JsonModel from './JsonModel'

export const undefToNull = data => {
	if (data === undefined || data === null) {
		return null
	}
	if (typeof data !== 'object') {
		return data
	}
	if (Array.isArray(data)) {
		return data.map(undefToNull)
	}
	const out = {...data}
	Object.entries(out).forEach(([key, value]) => {
		out[key] = undefToNull(value)
	})
	return out
}

export const getId = async (model, data) => {
	let id = data[model.idCol]
	if (id == null) {
		// Be sure to call with model as this, like in JsonModel
		id = await model.columns[model.idCol].value.call(model, data)
	}
	// This can only happen for integer ids
	if (id == null) id = await model.getNextId()
	return id
}

const REMOVE = 0
const SET = 1
const INSERT = 2
const UPDATE = 3
const SAVE = 4

class ESModel extends JsonModel {
	constructor({dispatch, ...options}) {
		super(options)
		this.dispatch = dispatch
		this.writable = false
	}

	TYPE = `es/${this.name}`

	setWritable(state) {
		// Slight hack: use the writable state to fall back to JsonModel behavior
		// This makes deriver and migrations work without changes
		// Note: during writable, no events are created. Be careful.
		this.writable = state
	}

	async set(obj, insertOnly) {
		if (this.writable) return super.set(obj, insertOnly)
		const {data} = await this.dispatch(this.TYPE, [
			insertOnly ? INSERT : SET,
			obj,
		])
		const id = data[2]
		// Note, his could return a later version of the object
		return this.get(id)
	}

	async update(o, upsert) {
		if (this.writable) return super.update(o, upsert)
		let id = o[this.idCol]
		if (id == null && !upsert) {
			throw new TypeError('No ID specified')
		}
		const {data} = await this.dispatch(this.TYPE, [
			upsert ? SAVE : UPDATE,
			undefToNull(o),
		])
		id = data[2]
		// Note, his could return a later version of the object
		return this.get(id)
	}

	updateNoTrans(obj, upsert) {
		if (this.writable) return super.updateNoTrans(obj, upsert)
		throw new Error('Non-transactional changes are not possible with ESModel')
	}

	async remove(idOrObj) {
		if (this.writable) return super.remove(idOrObj)
		const id = typeof idOrObj === 'object' ? idOrObj[this.idCol] : idOrObj
		if (id == null) throw new TypeError('No ID specified')
		await this.dispatch(this.TYPE, [REMOVE, id])
		return true
	}

	changeId() {
		throw new Error(`ESModel doesn't support changeId yet`)
	}

	_maxId = 0

	async getNextId() {
		if (!this._maxId) this._maxId = await this.max(this.idCol)
		return ++this._maxId
	}

	async applyChanges(result) {
		this._maxId = 0
		return super.applyChanges(result)
	}

	static async preprocessor({model, event}) {
		if (event.type !== model.TYPE) return
		if (event.data[0] > REMOVE) {
			// Always overwrite, so repeat events get correct ids
			event.data[2] = await getId(model, event.data[1])
			return event
		}
	}

	static async reducer(model, {type, data}) {
		if (!model || type !== model.TYPE) return false

		let [action, obj, id] = data
		if (action === REMOVE) {
			if (await model.exists({[model.idCol]: obj})) return {rm: [obj]}
			return false
		}

		if (obj[model.idCol] == null) obj = {...obj, [model.idCol]: id}

		const exists = await model.exists({[model.idCol]: id})

		switch (action) {
			case SET:
				return exists ? {set: [obj]} : {ins: [obj]}
			case INSERT:
				return exists ? {error: `object ${id} already exists`} : {ins: [obj]}
			case UPDATE:
				return exists ? {upd: [obj]} : {error: `object ${id} does not exist`}
			case SAVE:
				return exists ? {upd: [obj]} : {ins: [obj]}
			default:
				throw new TypeError('db action not found')
		}
	}
}

export default ESModel
