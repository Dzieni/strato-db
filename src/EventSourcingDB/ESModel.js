// Drop-in replacement for JsonModel
// Caveats:
// * `.update()` returns the current object at the time of returning, not the one that was updated
//
// Events all type `es/name` and data `[actionEnum, id, obj]`
// The id is assigned by the preprocessor except for RM

import JsonModel from '../JsonModel'
import {DEV} from '../lib/warning'

export const undefToNull = data => {
	if (data == null) return null
	if (typeof data !== 'object') return data
	if (Array.isArray(data)) return data.map(undefToNull)
	if (Object.getPrototypeOf(data) !== Object.prototype) return data
	const out = {}
	Object.entries(data).forEach(([key, value]) => {
		out[key] = undefToNull(value)
	})
	return out
}

export const getId = async (model, data) => {
	let id = data[model.idCol]
	if (id == null) {
		// Be sure to call with model as this, like in JsonModel
		id = await model.columns[model.idCol].value.call(model, data)
	}
	// This can only happen for integer ids
	if (id == null) id = await model.getNextId()
	return id
}

class ESModel extends JsonModel {
	/* eslint-disable lines-between-class-members */
	static REMOVE = 0
	static SET = 1
	static INSERT = 2
	static UPDATE = 3
	static SAVE = 4
	/* eslint-enable lines-between-class-members */

	constructor({dispatch, ...options}) {
		super(options)
		this.dispatch = dispatch
		this.writable = false
	}

	TYPE = `es/${this.name}`

	setWritable(state) {
		// Slight hack: use the writable state to fall back to JsonModel behavior
		// This makes deriver and migrations work without changes
		// Note: during writable, no events are created. Be careful.
		this.writable = state
	}

	async set(obj, insertOnly, noReturn, meta) {
		if (DEV && noReturn != null && typeof noReturn !== 'boolean')
			throw new Error(`${this.name}: meta argument is now in fourth position`)
		if (this.writable) return super.set(obj, insertOnly, noReturn)

		const d = [insertOnly ? ESModel.INSERT : ESModel.SET, null, obj]
		if (meta) d[3] = meta

		const {
			data,
			// We know this is always present, our reducer always
			result: {[this.name]: r},
		} = await this.dispatch(this.TYPE, d)
		const id = data[1]

		if (r.esFail) throw new Error(`${this.name}.set ${id}: ${r.esFail}`)

		// We have to get because we don't know what calculated values did
		// Unfortunately, this might be the object after a later event
		return noReturn ? undefined : this.get(id)
	}

	async update(o, upsert, noReturn, meta) {
		if (DEV && noReturn != null && typeof noReturn !== 'boolean')
			throw new Error(`${this.name}: meta argument is now in fourth position`)

		if (this.writable) return super.update(o, upsert, noReturn)

		if (DEV && noReturn != null && typeof noReturn !== 'boolean')
			throw new Error(`${this.name}: meta argument is now in fourth position`)
		let id = o[this.idCol]
		if (id == null && !upsert) throw new TypeError('No ID specified')

		const d = [upsert ? ESModel.SAVE : ESModel.UPDATE, null, undefToNull(o)]
		if (meta) d.push(meta)

		const {
			data,
			result: {[this.name]: r},
		} = await this.dispatch(this.TYPE, d)
		id = data[1]

		if (r.esFail) throw new Error(`${this.name}.update ${id}: ${r.esFail}`)

		// We have to get because we don't know what calculated values did
		// Unfortunately, this might be the object after a later event
		return this.get(id)
	}

	updateNoTrans(obj, upsert) {
		if (this.writable) return super.updateNoTrans(obj, upsert)
		throw new Error('Non-transactional changes are not possible with ESModel')
	}

	async remove(idOrObj, meta) {
		if (this.writable) return super.remove(idOrObj)
		const id = typeof idOrObj === 'object' ? idOrObj[this.idCol] : idOrObj
		if (id == null) throw new TypeError('No ID specified')

		const d = [ESModel.REMOVE, id]
		if (meta) d[3] = meta

		await this.dispatch(this.TYPE, d)
		return true
	}

	changeId() {
		throw new Error(`ESModel doesn't support changeId yet`)
	}

	_maxId = 0

	async getNextId() {
		if (!this._maxId) this._maxId = await this.max(this.idCol)
		return ++this._maxId
	}

	async applyChanges(result) {
		if (result.esFail) return
		this._maxId = 0
		return super.applyChanges({...result, esFail: undefined})
	}

	static async preprocessor({model, event}) {
		if (event.type !== model.TYPE) return
		if (event.data[0] > ESModel.REMOVE) {
			// Always overwrite, so repeat events get correct ids
			event.data[1] = await getId(model, event.data[2])
			return event
		}
	}

	static async reducer(model, {type, data}) {
		if (!model || type !== model.TYPE) return false

		let [action, id, obj] = data
		if (action === ESModel.REMOVE) {
			if (await model.exists({[model.idCol]: id})) return {rm: [id]}
			return false
		}

		if (obj[model.idCol] == null) obj = {...obj, [model.idCol]: id}

		const exists = await model.exists({[model.idCol]: id})

		switch (action) {
			case ESModel.SET:
				return exists ? {set: [obj]} : {ins: [obj]}
			case ESModel.INSERT:
				return exists ? {esFail: 'EEXIST'} : {ins: [obj]}
			case ESModel.UPDATE:
				return exists ? {upd: [obj]} : {esFail: 'ENOENT'}
			case ESModel.SAVE:
				return exists ? {upd: [obj]} : {ins: [obj]}
			default:
				throw new TypeError('db action not found')
		}
	}
}

export default ESModel
