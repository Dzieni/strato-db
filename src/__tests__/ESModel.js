import ESModel, {undefToNull, getId} from '../ESModel'
import {withESDB} from './_helpers'

class ESModelCustomId extends ESModel {
	constructor(options) {
		super({
			...options,
			columns: {
				myId: {
					index: true,
				},
			},
			idCol: 'myId',
		})
	}
}

class ESModelIntId extends ESModel {
	constructor(options) {
		super({
			...options,
			columns: {
				myId: {
					index: true,
					type: 'INTEGER',
				},
			},
			idCol: 'myId',
		})
	}
}

const sampleObject = {id: 'asd', meep: 'moop', top: 'kek'}

test('undefToNull', () => {
	const obj = {
		a: 'b',
		c: 4,
		nested: {
			stuff: 'test',
			asd: undefined,
			dsa: null,
		},
		array: [
			'two plus two is',
			undefined,
			{minus: 1, thats: '?'},
			null,
			'quick maths',
		],
		zxc: null,
		cxz: undefined,
	}
	expect(undefToNull(obj)).toEqual({
		...obj,
		nested: {...obj.nested, asd: null},
		array: [
			'two plus two is',
			null,
			{minus: 1, thats: '?'},
			null,
			'quick maths',
		],
		cxz: null,
	})
})

test('create', () =>
	withESDB(
		async eSDB => {
			const result = await eSDB.db.models.test.all()
			expect(result).toEqual([])
		},
		{test: {Model: ESModel}}
	))

test('getId standard settings', () =>
	withESDB(
		async eSDB => {
			expect(await getId(eSDB.db.models.test, {top: 'kek'})).toBeTruthy()
		},
		{test: {Model: ESModel}}
	))

test('getId custom idCol', () =>
	withESDB(
		async eSDB => {
			expect(await getId(eSDB.db.models.test, {top: 'kek'})).toBeTruthy()
		},
		{test: {Model: ESModelCustomId}}
	))

test('getId integer idCol', () =>
	withESDB(
		async eSDB => {
			expect(await getId(eSDB.db.models.test, {top: 'kek'})).toBe(1)
			expect(await getId(eSDB.db.models.test, {top: 'kek'})).toBe(2)
			await eSDB.db.models.test.set({myId: 7, asd: 'dsa'})
			expect(await getId(eSDB.db.models.test, {moop: 'meep'})).toBe(8)
		},
		{
			test: {Model: ESModelIntId},
		}
	))

test('set w/ id', () =>
	withESDB(
		async eSDB => {
			expect(await eSDB.db.models.test.set(sampleObject)).toEqual(sampleObject)
		},
		{test: {Model: ESModel}}
	))

test('set w/o id', () =>
	withESDB(
		async eSDB => {
			const newObject = await eSDB.db.models.test.set({
				...sampleObject,
				id: undefined,
			})
			expect(await eSDB.db.models.test.get(newObject.id)).toEqual(newObject)
		},
		{test: {Model: ESModel}}
	))

test('set w/o id int', () =>
	withESDB(
		async eSDB => {
			const sampleWithoutId = {
				sampleObject,
				id: undefined,
			}
			await Promise.all([
				eSDB.db.models.test.set(sampleWithoutId),
				eSDB.db.models.test.set(sampleWithoutId),
				eSDB.db.models.test.set(sampleWithoutId),
			])
			expect(await getId(eSDB.db.models.test, {top: 'kek'})).toEqual(4)
			expect(await eSDB.db.models.test.count()).toEqual(3)
		},
		{test: {Model: ESModelIntId}}
	))

test('set insertOnly', () =>
	withESDB(
		async eSDB => {
			await expect(
				Promise.all([
					eSDB.db.models.test.set(sampleObject, true),
					eSDB.db.models.test.set(sampleObject, true),
				])
			).rejects.toHaveProperty('error.test')
		},
		{test: {Model: ESModel}}
	))

test('update', () =>
	withESDB(
		async eSDB => {
			await eSDB.db.models.test.set(sampleObject)
			const newObject = {
				id: sampleObject.id,
				top: 'bottom',
				new: 'prop',
			}
			await eSDB.db.models.test.update(newObject)
			const items = await eSDB.db.models.test.all()
			expect(items).toHaveLength(1)
			expect(items[0]).toEqual({...sampleObject, ...newObject})
		},
		{test: {Model: ESModel}}
	))

test('update w/o id', () =>
	withESDB(
		async eSDB => {
			await expect(eSDB.db.models.test.update({top: 'kek'})).rejects.toThrow(
				'No ID specified'
			)
		},
		{test: {Model: ESModel}}
	))

test('update w/ undefined values', () =>
	withESDB(
		async eSDB => {
			await eSDB.db.models.test.set(sampleObject)
			expect(
				await eSDB.db.models.test.update({id: sampleObject.id, top: undefined})
			).toEqual({...sampleObject, top: null}) // we cannot pass undefined here
		},
		{test: {Model: ESModel}}
	))

test('update non-existent object', () =>
	withESDB(
		async eSDB => {
			await expect(
				eSDB.db.models.test.update(sampleObject)
			).rejects.toHaveProperty('error.test')
		},
		{test: {Model: ESModel}}
	))

test('update upsert', () =>
	withESDB(
		async eSDB => {
			expect(await eSDB.db.models.test.update(sampleObject, true)).toEqual(
				sampleObject
			)
		},
		{test: {Model: ESModel}}
	))

test('update upsert w/o id', () =>
	withESDB(
		async eSDB => {
			expect(await eSDB.db.models.test.update({top: 'kek'}, true)).toEqual({
				myId: 1,
				top: 'kek',
			})
		},
		{test: {Model: ESModelIntId}}
	))

test('remove by id', () =>
	withESDB(
		async eSDB => {
			await eSDB.db.models.test.set(sampleObject)
			expect(await eSDB.db.models.test.remove(sampleObject.id)).toEqual(true)
			expect(await eSDB.db.models.test.all()).toEqual([])
		},
		{test: {Model: ESModel}}
	))

test('remove by object', () =>
	withESDB(
		async eSDB => {
			await eSDB.db.models.test.set(sampleObject)
			expect(await eSDB.db.models.test.remove(sampleObject)).toEqual(true)
			expect(await eSDB.db.models.test.all()).toEqual([])
		},
		{test: {Model: ESModel}}
	))

test('remove by object without id', () =>
	withESDB(
		async eSDB => {
			await eSDB.db.models.test.set(sampleObject)
			await expect(
				eSDB.db.models.test.remove({...sampleObject, id: undefined})
			).rejects.toThrow()
		},
		{test: {Model: ESModel}}
	))

class Foo {
	getF() {
		return this.f
	}
}

test('ItemClass set', async () =>
	withESDB(
		async eSDB => {
			const {m} = eSDB.store
			const setted = await m.set({id: 4, f: 'meep'})
			expect(setted instanceof Foo).toBe(true)
			expect(setted.getF()).toBe('meep')
		},
		{
			m: {
				Model: class extends ESModel {
					constructor(o) {
						super({...o, ItemClass: Foo})
					}
				},
			},
		}
	))
