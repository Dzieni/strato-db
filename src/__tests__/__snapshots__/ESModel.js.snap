// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`events 1`] = `
Array [
  Object {
    "data": Array [
      1,
      Object {
        "meep": "moop",
      },
      1,
    ],
    "result": Object {
      "m": Object {
        "ins": Array [
          Object {
            "id": 1,
            "meep": "moop",
          },
        ],
      },
    },
    "ts": 0,
    "type": "es/m",
    "v": 1,
  },
  Object {
    "data": Array [
      3,
      Object {
        "beep": "boop",
        "id": 1,
      },
      1,
    ],
    "result": Object {
      "m": Object {
        "upd": Array [
          Object {
            "beep": "boop",
            "id": 1,
          },
        ],
      },
    },
    "ts": 0,
    "type": "es/m",
    "v": 2,
  },
  Object {
    "data": Array [
      2,
      Object {
        "meep": "moop",
      },
      2,
    ],
    "result": Object {
      "m": Object {
        "ins": Array [
          Object {
            "id": 2,
            "meep": "moop",
          },
        ],
      },
    },
    "ts": 0,
    "type": "es/m",
    "v": 3,
  },
  Object {
    "data": Array [
      4,
      Object {
        "beep": "boop",
        "id": 3,
      },
      3,
    ],
    "result": Object {
      "m": Object {
        "ins": Array [
          Object {
            "beep": "boop",
            "id": 3,
          },
        ],
      },
    },
    "ts": 0,
    "type": "es/m",
    "v": 4,
  },
  Object {
    "data": Array [
      0,
      3,
    ],
    "result": Object {
      "m": Object {
        "rm": Array [
          3,
        ],
      },
    },
    "ts": 0,
    "type": "es/m",
    "v": 5,
  },
]
`;
